src/dmd/dsymbol.d

```diff
@@ -2078,44 +2077,40 @@ extern (C++) final class ForwardingScopeDsymbol : ScopeDsymbol
  */
 extern (C++) final class DsymbolTable : RootObject
 {
-    AA* tab;
+    Dsymbol[void*] tab;
 
     // Look up Identifier. Return Dsymbol if found, NULL if not.
     Dsymbol lookup(const Identifier ident)
     {
         //printf("DsymbolTable::lookup(%s)\n", (char*)ident.string);
-        return cast(Dsymbol)dmd_aaGetRvalue(tab, cast(void*)ident);
+        auto ps = cast(void*)ident in tab;
+        return ps ? *ps : null;
     }
 
     // Insert Dsymbol in table. Return NULL if already there.
     Dsymbol insert(Dsymbol s)
     {
         //printf("DsymbolTable::insert(this = %p, '%s')\n", this, s.ident.toChars());
-        const ident = s.ident;
-        Dsymbol* ps = cast(Dsymbol*)dmd_aaGet(&tab, cast(void*)ident);
-        if (*ps)
+        if (cast(void*)s.ident in tab)
             return null; // already in table
-        *ps = s;
+        tab[cast(void*)s.ident] = s;
```

Comment:
The trouble here is this adds a second lookup. This is not acceptable (performance reasons).

Reply:
You were happy to add one to #5428.

Reply:
Yes, and that wasn't in a hot section of code. The dmd symbol table is very hot.

Reply:
Well in the absence of a getOrCreate property for built-in AAs, not much can really be done to help that there.

Reply:
Well, there is _aaGetY which could be called directly, but I'm not sure we should go that road. It is already used by `dup`, maybe a similar small helper in object.d would be a nice extension. IIRC there are also a couple of enhancement requests in bugzilla for this.

Reply:
Well, `_aaGetY` does not have enough information to say whether or not a value existed before a lookup was requested.  One would have to either define a new internal function, move the common bits into new helpers, and add the difference

i.e:
```
void* _aaGetY(...)
{
  auto p = aaLookup(...);
  if (!p)
    p = aaInsert(...);
  return p;
}

void* _aaGetOrInsert(..., ref bool existed)
{
  existed = false;
  auto p = aaLookup(...);
  if (!p)
    p = aaInsert(...);
  else
    existed = true;
  return p;
}
```

Reply:
In general the information `existed` is desirable, but for dmd, _aaGetY has the same semantics as dmd_aaGet. The zero initialization is used to detect whether the key has been added.